File: frontend/propmate-web/src/services/favouriteService.js
const API_URL = "http://localhost:5235/api/favourites";

// Add a property to favourites
export async function addFavourite(propertyId, userId) {
    const response = await fetch(API_URL, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify({
            propertyId,
            userId,
        }),
    });

    if (!response.ok) {
        if (response.status === 409) {
            throw new Error("Property is already in favourites.");
        }

        throw new Error("Failed to add favourite.");
    }

    return await response.json();
}

// Remove a favourite
export async function removeFavourite(favouriteId) {
    const response = await fetch(`${API_URL}/${favouriteId}`, {
        method: "DELETE",
    });

    if (!response.ok) {
        throw new Error("Failed to remove favourite.");
    }
}

File: mobile/propmate_mobile/lib/services/favourite_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models/favourite.dart';

class FavouriteService {
  static const String baseUrl = 'http://localhost:5235/api/favourites';

  Future<Favourite> addFavourite({
    required int propertyId,
    required int userId,
  }) async {
    final response = await http.post(
      Uri.parse(baseUrl),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'propertyId': propertyId, 'userId': userId}),
    );

    if (response.statusCode == 201) {
      return Favourite.fromJson(jsonDecode(response.body));
    }

    if (response.statusCode == 409) {
      throw Exception('Property is already in favourites');
    }

    if (response.statusCode == 404) {
      throw Exception('Property not found');
    }

    throw Exception('Failed to add favourite');
  }

  Future<void> removeFavourite(int favouriteId) async {
    final response = await http.delete(Uri.parse('$baseUrl/$favouriteId'));

    if (response.statusCode == 204) {
      return;
    }

    if (response.statusCode == 404) {
      throw Exception('Favourite not found');
    }

    throw Exception('Failed to remove favourite');
  }
}
